### ISSUE TITLE
Setting PNG image metadata doesn't work

### ISSUE THREAD

**Issue by Evheni** [dotnet/wpf#7121]
Code example doesn't work. TrySave method is called before the changes in metadata.
If we add saving after the changes, the TrySave method returns false.
<img width="1259" alt="image" src="https://user-images.githubusercontent.com/15703832/192312863-29764519-bf26-4230-97a4-5f04e93f9e21.png">

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 476ff378-df63-0a54-8a1b-4b8c9dc54c67
* Version Independent ID: f8e5a0e6-da8f-ccab-8dec-7d7cbe00c497
* Content: [How to: Write Metadata to a Bitmap - WPF .NET Framework](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/graphics-multimedia/how-to-write-metadata-to-a-bitmap?view=netframeworkdesktop-4.8)
* Content Source: [dotnet-desktop-guide/framework/wpf/graphics-multimedia/how-to-write-metadata-to-a-bitmap.md](https://github.com/dotnet/docs-desktop/blob/main/dotnet-desktop-guide/framework/wpf/graphics-multimedia/how-to-write-metadata-to-a-bitmap.md)
* Product: **dotnet-framework**
* Technology: **dotnet-wpf**
* GitHub Login: @adegeo
* Microsoft Alias: **adegeo**

**Comment by adegeo** [comment]
@Evheni I've transferred this issue to the WPF team.

I can't figure this out. I've been trying to for 1.5 hours and it just doesn't work. Can someone please help this customer?

**Comment by miloush** [comment]
You most certainly need to call `TrySave` after `SetQuery` which is the only way to invoke `IWICFastMeatadataEncoder::Commit`, see an example here: https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nn-wincodec-iwicfastmetadataencoder

If it returns `false`, which it does in this case as already hinted by the description above, you need to reencode. 

I am not sure how the example in docs turned into using PNG file, the documentation for `TrySave` suggests it was supposed to be a TIFF file: https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.imaging.inplacebitmapmetadatawriter.trysave?view=windowsdesktop-6.0

**Comment by adegeo** [comment]
After a lot of trial and error, I got it working:

```csharp
const string FILE = @"c:\path\to\your\file.png";

// Open the file for reading
using (Stream pngStream = new FileStream(FILE, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
{
    var pngDecoder = new PngBitmapDecoder(pngStream, BitmapCreateOptions.PreservePixelFormat, BitmapCacheOption.Default);
    BitmapFrame pngFrame = pngDecoder.Frames[0];
    InPlaceBitmapMetadataWriter pngInplace = pngFrame.CreateInPlaceBitmapMetadataWriter();

    // Set metadata here
    pngInplace.SetQuery("/Text/Description", "Have a nice day.");

    // Try to save in place, if failed, recreate the png
    if (!pngInplace.TrySave())
    {
        // Use a secondary stream for writing
        using (var pngSaveStream = new FileStream(FILE, FileMode.Open, FileAccess.Write, FileShare.ReadWrite))
        {
            var pngEncoder = new PngBitmapEncoder();

            // Duplicate each frame
            foreach (var frame in pngDecoder.Frames)
                pngEncoder.Frames.Add(BitmapFrame.Create(frame));

            // Set the required metadata
            ((BitmapMetadata)pngEncoder.Frames[0].Metadata).SetQuery("/Text/Description", "Have a nice day.");

            pngEncoder.Save(pngSaveStream);
        } // close\dispose pngSaveStream
    }
} // close\dispose pngStream
```

**Comment by Evheni** [comment]
Works for me. Thanks.
Will you update the example section?

**Comment by miloush** [comment]
I agree the documentation should be fixed, but this is not a very nice example. It also does not make much sense, since the built-in codec does not support in-place metadata for PNG at all, see the table here: https://learn.microsoft.com/en-us/windows/win32/wic/-wic-about-metadata#supported-metadata-formats

**Comment by adegeo** [comment]
Sure thing, I'll get things updated.

**Comment by adegeo** [comment]
@miloush I tried using a jpg but I get an error trying to save the file:

> **System.NotSupportedException:** 'No imaging component suitable to complete this operation was found.'

```csharp
using (var imageStream = new System.IO.FileStream("smiley.jpg", FileMode.Open,
                                                                FileAccess.ReadWrite,
                                                                FileShare.ReadWrite))
{
    var decoder = new JpegBitmapDecoder(imageStream,
                                            BitmapCreateOptions.PreservePixelFormat,
                                            BitmapCacheOption.Default);

    BitmapFrame frame = decoder.Frames[0];

    // Try to set the metadata directly on the image
    InPlaceBitmapMetadataWriter inplaceMetadata = frame.CreateInPlaceBitmapMetadataWriter();
    inplaceMetadata.SetQuery("/Text/Description", "Have a nice day.");

    if (!inplaceMetadata.TrySave())
    {
        // In place metadata save failed, must rencode image
        using (var saveStream = new FileStream("smiley.jpg", FileMode.Open,
                                                                FileAccess.Write,
                                                                FileShare.ReadWrite))
        {
            var encoder = new JpegBitmapEncoder();

            // Duplicate each frame
            foreach (var cloneFrame in decoder.Frames)
                encoder.Frames.Add(BitmapFrame.Create(cloneFrame));

            // Set metadata
            ((BitmapMetadata)encoder.Frames[0].Metadata).SetQuery("/Text/Description", "Have a nice day.");

            encoder.Save(saveStream); // <------- ERROR
        }
    }
}
```

**Comment by miloush** [comment]
How did you come up with `/Text/Description`? I believe that is supposed to be PNG tEXt chunk. Would e.g. `System.Title` work for you? See https://learn.microsoft.com/en-us/windows/win32/wic/-wic-native-image-format-metadata-queries for supported format-specific queries.

**Comment by adegeo** [comment]
I don't know, it was in the example. The error was vague so I didn't really know what to do. Thank you for that link, it clears up a lot of things.

**Comment by miloush** [comment]
Yeah just scratch that example. On one of the pages it even says it's an example for TIFF files...

I think it's important for users to know that WIC is providing the imaging features in WPF and they can reach out there for details. If you can sneak that somewhere somehow, that would be great!

**Comment by adegeo** [comment]
Yep, will do!

**Comment by adegeo** [comment]
This is still in progress.. I've just not found the time to get back to the docs and update them.

**Comment by ThomasGoulet73** [comment]
@adegeo Is this still in progress ? If so, I think we can remove the  waiting-author-feedback label.
